blob/azureblob/azureblob.go

```diff
@@ -108,105 +114,111 @@ const (
-	if err != nil {
-		return nil, err
+		AccountName AccountName
+		AccountKey  AccountKey
+		SASToken    SASToken
 	}
-
-	err = json.Unmarshal(f, ac)
-	if err != nil {
-		return nil, err
-	}
-
-	if ac.AccountKey != "" {
-		serviceURL, err := ServiceURLFromAccountKey(ac.AccountName, ac.AccountKey)
+	ac := AzureCreds{}
+	if credPath := u.Query()["cred_path"]; len(credPath) > 0 {
+		f, err := ioutil.ReadFile(credPath[0])
 		if err != nil {
 			return nil, err
 		}
-
-		credential, err := azblob.NewSharedKeyCredential(ac.AccountName, ac.AccountKey)
+		err = json.Unmarshal(f, &ac)
 		if err != nil {
 			return nil, err
 		}
-
-		opts.Credential = credential
-		return openBucket(ctx, serviceURL, u.Host, opts)
+	} else {
+		// Use default credential info from the environment.
+		ac.AccountKey = DefaultAccountKey()
+		ac.AccountName = DefaultAccountName()
+		ac.SASToken = DefaultSASToken()
 	}
 
-	serviceURL, err := ServiceURLFromSASToken(ac.AccountName, ac.SASToken)
-	if err != nil {
-		return nil, err
+	// azblob.Credential is an interface; we will use either a SharedKeyCredential
+	// or anonymous credentials. If the former, we will also fill in
+	// Options.Credential so that SignedURL will work.
+	var credential azblob.Credential
+	var sharedKeyCred *azblob.SharedKeyCredential
+	if ac.AccountKey != "" {
+		var err error
+		sharedKeyCred, err = NewCredential(ac.AccountName, ac.AccountKey)
+		if err != nil {
+			return nil, err
+		}
+		credential = sharedKeyCred
+	} else {
+		credential = azblob.NewAnonymousCredential()
 	}
-	return openBucket(ctx, serviceURL, u.Host, opts)
+	pipeline := NewPipeline(credential, azblob.PipelineOptions{})
+	return openBucket(ctx, pipeline, ac.AccountName, u.Host, &Options{
+		Credential: sharedKeyCred,
+		SASToken:   ac.SASToken,
+	})
+}
+
+// DefaultIdentity is a Wire provider set that provides an Azure storage
+// account name, key, and SharedKeyCredential from environment variables.
+var DefaultIdentity = wire.NewSet(
+	DefaultAccountName,
+	DefaultAccountKey,
+	NewCredential,
+	wire.Bind(new(azblob.Credential), new(azblob.SharedKeyCredential)),
+	wire.Value(azblob.PipelineOptions{}),
+)
+
+// SASTokenIdentity is a Wire provider set that provides an Azure storage
+// account name, SASToken, and anonymous credential from environment variables.
+var SASTokenIdentity = wire.NewSet(
+	DefaultAccountName,
+	DefaultSASToken,
+	azblob.NewAnonymousCredential,
+	wire.Value(azblob.PipelineOptions{}),
+)
+
+// AccountName is an Azure storage account name.
+type AccountName string
+
+// AccountKey is an Azure storage account key (primary or secondary).
+type AccountKey string
+
+// SASToken is an Azure shared access signature.
+// https://docs.microsoft.com/en-us/azure/storage/common/storage-dotnet-shared-access-signature-part-1
+type SASToken string
+
+// DefaultAccountName loads the Azure storage account name from the
+// AZURE_STORAGE_ACCOUNT environment variable.
+func DefaultAccountName() AccountName {
```

Comment:
Perhaps these should return errors if their environment variables are empty?

Reply:
OK, done.